### PR TITLE
Restarting Cases

### PR DESCRIPTION
**Pull request by anshgupta1234**
This PR addresses #95 
- Adds docs for restarting cases
- Makes some changes to `pre_process` that allow for adding new patches while restarting
- New example case, 2D_shockbubble_restart

### CODE REVIEW

**Comment by sbryngelson** [comment]
After running `case.py` and then running `./mfc.sh run -n 8 ./examples/2D_shockbubble_restart/restart_case.py -t pre_process`

I get 
```console
 Pre-processing a 300x100x0 case on 8 rank(s)
mkdir: ./restart_data/lustre_3848.dat: File exists
 Final Time   2.7792999999999998E-002

    Done (in 0:00:00.268480)
mfc: (venv) Exiting the Python virtual environment.
```

Something is going wrong here: `mkdir: ./restart_data/lustre_3848.dat: File exists`

**Comment by sbryngelson** [comment]
I also notice that this case is very long. I would rather have a simple case that can be executed and shown to work in 1D, even, is fine, in just a moment or two. This simulation takes ~30 minutes for me.

**Comment by sbryngelson** [comment]
The `running.md` file has syntax errors. For example, it isn't showing code variables `like_this`. Also, lustre files are only output for certain `case.py` parameters. What if one is using binary files?

File: docs/documentation/running.md [hunk 1]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+- Run pre-process and simulation on the case.
+    - `./mfc.sh run case.py -t pre_process simulation `
+- As the simulation runs, it will create LUSTRE files for each saved timestep in `./restart_data`.
+- When the simulation stops, choose any LUSTRE file as the restarting point (lustre_ $t_s$.dat)
+- Create a new duplicate input file, (ex. restart_case.py), on which it should:
+
+1. For the Computational Domain Parameters 
+    - Have the following removed BUT m, n, and p:
+		- All domaing/mesh information
+			- (xyz)_domain%beg
```
Comment (sbryngelson) [inline docs/documentation/running.md]: these should be shown like `this`

File: docs/documentation/running.md [hunk 2]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+			- (xyz)_domain%end
+			- stretch_(xyz)
+			- a_(xyz)
+			- (xyz)_a
+			- (xyz)_b
+	- Have the following altered:
+		- t_step_start : $t_s$ # this is the point at which the simulation will restart
```
Comment (sbryngelson) [inline docs/documentation/running.md]: use `this` formatting

File: docs/documentation/running.md [hunk 3]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+		- t_step_stop  : $t_{f2}$ # your new final simulation time, which can be the same as $t_f$
+		- t_step_save  : ${SF}_2$ # if interested in changing the saving frequency 
+	- Have the following ADDED:
+		- old_ic : 'T' # to specify that we have initial conditions from previous simulations
+		- old_grid : 'T' # to specify that we have a grid from previous simulations (maybe I do not need m, n, and p, then?)
+		- t_step_old : $t_i$ # this is the time step used as the t_start of the original case.py file
+2. For the Simulation Algorithm Parameters
+	- Substitute 'num_patches' to reflect the number of ADDED patches in the restart_case.py. If no patches are added, set 'num_patches = 0'
```
Comment (sbryngelson) [inline docs/documentation/running.md]: Please avoid all caps.

File: docs/documentation/running.md [hunk 4]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+
+3. For Patches
+	- Have all information about old patches (used in the case.py file) removed.
```
Comment (sbryngelson) [inline docs/documentation/running.md]: should be `case.py`.

File: docs/documentation/running.md [hunk 5]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+		- patch_icpp(1)%all variables
```
Comment (sbryngelson) [inline docs/documentation/running.md]: fix

File: docs/documentation/running.md [hunk 6]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+		- patch_icpp(2)%all variables
+		- patch_icpp(num_patches)%all variables
+	- Add information about new patches that will be introduced, if any. The parameter num_patches should reflect this addition.
+		- e.g. patch_icpp(1)%some variables of interest
```
Comment (sbryngelson) [inline docs/documentation/running.md]: fix

File: docs/documentation/running.md [hunk 7]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+
+3. For Patches
+	- Have all information about old patches (used in the case.py file) removed.
+		- patch_icpp(1)%all variables
+		- patch_icpp(2)%all variables
+		- patch_icpp(num_patches)%all variables
+	- Add information about new patches that will be introduced, if any. The parameter num_patches should reflect this addition.
+		- e.g. patch_icpp(1)%some variables of interest	
+
+4. For Fluid properties
```
Comment (sbryngelson) [inline docs/documentation/running.md:139]: unclear why this item is necessary.

File: docs/documentation/running.md [hunk 8]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+$ ./mfc.sh run examples/2D_shockbubble_restart/case.py -t pre_process simulation
+$ ./mfc.sh run examples/2D_shockbubble_restart/restart_case.py -t pre_process simulation
+$ ./mfc.sh run examples/2D_shockbubble_restart/case.py -t post_process
+$ ./mfc.sh run examples/2D_shockbubble_restart/restart_case.py -t post_process
+```
+
+The result will look like this:
+![](../res/patches.gif)
+
+In the 3rd frame, we can see the added patch!
```
Comment (sbryngelson) [inline docs/documentation/running.md]: remove "!" -> "."

File: docs/documentation/running.md [hunk 9]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+
+- Set up the initial simulation, with:
+    - t_step_start : $t_i$
+    - t_step_stop  : $t_f$
+    - t_step_save  : $SF$
+in which $t_i$ is the starting time, $t_f$ is the final time, and $SF$ is the saving frequency time.
+- Run pre-process and simulation on the case.
+    - `./mfc.sh run case.py -t pre_process simulation `
+- As the simulation runs, it will create LUSTRE files for each saved timestep in `./restart_data`.
+- When the simulation stops, choose any LUSTRE file as the restarting point (lustre_ $t_s$.dat)
```
Comment (sbryngelson) [inline docs/documentation/running.md:108]: what if not using lustre?

File: docs/documentation/running.md [hunk 10]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+- Create a new duplicate input file, (ex. restart_case.py), on which it should:
```
Comment (sbryngelson) [inline docs/documentation/running.md]: use `this.py` notation

File: docs/documentation/running.md [hunk 11]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+
+1. For the Computational Domain Parameters 
+    - Have the following removed BUT m, n, and p:
```
Comment (sbryngelson) [inline docs/documentation/running.md]: no caps, use `m` notation.

File: docs/documentation/running.md [hunk 12]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+			- a_(xyz)
+			- (xyz)_a
+			- (xyz)_b
+	- Have the following altered:
+		- t_step_start : $t_s$ # this is the point at which the simulation will restart
+		- t_step_stop  : $t_{f2}$ # your new final simulation time, which can be the same as $t_f$
+		- t_step_save  : ${SF}_2$ # if interested in changing the saving frequency 
+	- Have the following ADDED:
+		- old_ic : 'T' # to specify that we have initial conditions from previous simulations
+		- old_grid : 'T' # to specify that we have a grid from previous simulations (maybe I do not need m, n, and p, then?)
```
Comment (sbryngelson) [inline docs/documentation/running.md]: looks like you left something here... " (maybe I do not need m, n, and p, then?)"

File: docs/documentation/running.md [hunk 13]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+If you want to restart a simulation, 
+
+- Set up the initial simulation, with:
+    - t_step_start : $t_i$
+    - t_step_stop  : $t_f$
+    - t_step_save  : $SF$
+in which $t_i$ is the starting time, $t_f$ is the final time, and $SF$ is the saving frequency time.
+- Run pre-process and simulation on the case.
+    - `./mfc.sh run case.py -t pre_process simulation `
+- As the simulation runs, it will create LUSTRE files for each saved timestep in `./restart_data`.
```
Comment (sbryngelson) [inline docs/documentation/running.md:107]: i don't think this is always true

File: docs/documentation/running.md [hunk 14]
```diff
@@ -91,7 +91,81 @@ modified by users.
...
+### Restarting Cases
+
+When running a simulation, MFC generates a `./restart_data` folder in the case directory that contains `lustre_*.dat` files that can be used to restart a simulation from saved timesteps. This allows a user to run a simulation to some timestep $X$, then later continue it to run to another timestep $Y$, where $Y > X$. The user can also choose to add new patches at the intermediate timestep.
+
+If you want to restart a simulation, 
+
+- Set up the initial simulation, with:
+    - t_step_start : $t_i$
+    - t_step_stop  : $t_f$
+    - t_step_save  : $SF$
```
Comment (sbryngelson) [inline docs/documentation/running.md]: incorrect math notation here... $t_i$ would be a real number (presumably) and $SF$ would be $S$ multiplying $F$, whatever those variables are. consider switching to `t_i` and `SF` etc.

File: docs/documentation/running.md (hunk 10 above)
Comment (sbryngelson) [inline docs/documentation/running.md]: "on which it should" (grammar problem)

**Comment by anshgupta1234** [comment]
@sbryngelson I believe this is ready to merge.

**Review by sbryngelson — APPROVED** [review]

**Review by sbryngelson — APPROVED** [review]

**Pull request merged**